lib/storage/providers/SQLiteStorage.js

```diff
@@ -0,0 +1,118 @@
+
+// All of the 3 pragmas below were suggested by SQLite team.
+// You can find more info about them here: https://www.sqlite.org/pragma.html
+db.execute('PRAGMA CACHE_SIZE=-20000;');
+db.execute('PRAGMA synchronous=NORMAL;');
+db.execute('PRAGMA journal_mode=WAL;');
+
+const provider = {
+    /**
+    * Get the value of a given key or return `null` if it's not available in storage
+    * @param {String} key
+    * @return {Promise<*>}
+    */
+    getItem(key) {
+        return db.executeAsync('SELECT record_key, valueJSON FROM magic_map WHERE record_key = ?;', [key]).then(({rows}) => {
+            if (rows.length === 0) {
+                return null;
+            }
+            const result = rows.item(0);
+            return JSON.parse(result.valueJSON);
+        });
+    },
+
+    /**
+    * Get multiple key-value pairs for the given array of keys in a batch
+    * @param {String[]} keys
+    * @return {Promise<Array<[key, value]>>}
+    */
+    multiGet(keys) {
+        const placeholders = _.map(keys, () => '?').join(',');
+        const command = `SELECT record_key, valueJSON FROM magic_map WHERE record_key IN (${placeholders});`;
+        return db.executeAsync(command, keys)
+            .then(({rows}) => {
+                // eslint-disable-next-line no-underscore-dangle
+                const result = _.map(rows._array, row => [row.record_key, JSON.parse(row.valueJSON)]);
+                return result;
+            });
+    },
+
+    /**
+    * Sets the value for a given key. The only requirement is that the value should be serializable to JSON string
+    * @param {String} key
+    * @param {*} value
+    * @return {Promise<void>}
+    */
+    setItem(key, value) {
+        return db.executeAsync('REPLACE INTO magic_map (record_key, valueJSON) VALUES (?, ?);', [key, JSON.stringify(value)]);
+    },
+
+    /**
+    * Stores multiple key-value pairs in a batch
+    * @param {Array<[key, value]>} pairs
+    * @return {Promise<void>}
+    */
+    multiSet(pairs) {
+        const stringifiedPairs = _.map(pairs, pair => [
+            pair[0],
+            JSON.stringify(pair[1]),
+        ]);
+        return db.executeBatchAsync([['REPLACE INTO magic_map (record_key, valueJSON) VALUES (?, json(?));', stringifiedPairs]]);
+    },
+
+    /**
+    * Multiple merging of existing and new values in a batch
+    * @param {Array<[key, value]>} pairs
+    * @return {Promise<void>}
+    */
+    multiMerge(pairs) {
+        // Note: We use `ON CONFLICT DO UPDATE` here instead of `INSERT OR REPLACE INTO`
+        // so the new JSON value is merged into the old one if there's an existing value
+        const query = `INSERT INTO magic_map (record_key, valueJSON) 
+             VALUES (:key, JSON(:value)) 
+             ON CONFLICT DO UPDATE 
+             SET valueJSON = JSON_PATCH(valueJSON, JSON(:value)); 
+        `;
```

Comment:
Can you link to documentation regarding how exactly this works 

For example how are arrays merged - are they concatenated or merged like objects

Reply:
This looks like a good explanation here -> https://www.sqlite.org/json1.html#jpatch

Reply:
Ok then this a potentially breaking change due to array merging 
`JSON_PATCH` would not merge arrays but use the array from the input as the final array, while the old logic (and the one used on web is):

https://github.com/Expensify/react-native-onyx/blob/91aabb7219ddc8722cc2955de0398b1ae70c546d/lib/fastMerge.js#L24-L28

So old logic would do this

```js
const src = { list: [1, 2, 3, 4] };
const dest = { list: [4, 3] };
merge(src, dest); // { list: [4, 3, 3, 4] }
```

While the new

```js
const src = { list: [1, 2, 3, 4] };
const dest = { list: [4, 3] };
merge(src, dest); // { list: [4, 3] }
```

---

This `JSON_PATCH` simple logic might be for the best as it's easier to understand, but we should test whether it doesn't break anything (merging collections)

Reply:
I agree some quick testing (either manual or automated) would be good here. 

I recall that @iwiznia changed the behavior for array merging. And then we figured out later that it would be the same as the sqlite MergePatch behavior. That was before we implemented an even more custom merge function (`fastMerge()`). I would imagine that we would have tried to preserve that behavior, but don't know for sure.

Is it solved by this code here ?

https://github.com/Expensify/react-native-onyx/blob/91aabb7219ddc8722cc2955de0398b1ae70c546d/lib/fastMerge.js#L56-L64

Reply:
> Is it solved by this code here ?

Yes it is (I missed that part)
LocalForage provider seems to work the same way, though it uses `Object.assign`, so we should be good 

https://github.com/Expensify/react-native-onyx/blob/91aabb7219ddc8722cc2955de0398b1ae70c546d/lib/storage/providers/LocalForage.js#L28-L31